codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -0,0 +1,280 @@
+            // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
+            // The stream was already made active, but ctx may have been null so it wasn't initialized.
+            // https://github.com/netty/netty/issues/4942
+            new ConnectionListener().onStreamActive(stream);
+            upgrade.upgradeRequest().headers().setInt(
+                    HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
+            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
+                    .channelRead(ctx, upgrade.upgradeRequest().retain());
+        } finally {
+            upgrade.release();
+        }
+    }
+
+    // Override this to signal it will never throw an exception.
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        ctx.fireExceptionCaught(cause);
+    }
+
+    // Override this to signal it will never throw an exception.
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        ctx.flush();
+    }
+
+    /**
+     * Processes all {@link Http2Frame}s. {@link Http2StreamFrame}s may only originate in child
+     * streams.
+     */
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        if (!(msg instanceof Http2Frame)) {
+            ctx.write(msg, promise);
+            return;
+        }
+        try {
+            if (msg instanceof Http2WindowUpdateFrame) {
+                Http2WindowUpdateFrame frame = (Http2WindowUpdateFrame) msg;
+                consumeBytes(frame.streamId(), frame.windowSizeIncrement());
+            } else if (msg instanceof Http2StreamFrame) {
+                writeStreamFrame((Http2StreamFrame) msg, promise);
+            } else if (msg instanceof Http2GoAwayFrame) {
+                writeGoAwayFrame((Http2GoAwayFrame) msg, promise);
+            } else {
+                throw new UnsupportedMessageTypeException(msg);
+            }
+        } finally {
+            ReferenceCountUtil.release(msg);
+        }
+    }
+
+    private void consumeBytes(int streamId, int bytes) {
+        try {
+            Http2Stream stream = http2Handler.connection().stream(streamId);
+            http2Handler.connection().local().flowController()
+                        .consumeBytes(stream, bytes);
+        } catch (Throwable t) {
+            exceptionCaught(ctx, t);
+        }
+    }
+
+    private void writeGoAwayFrame(Http2GoAwayFrame frame, ChannelPromise promise) {
+        if (frame.lastStreamId() > -1) {
+            throw new IllegalArgumentException("Last stream id must not be set on GOAWAY frame");
+        }
+
+        int lastStreamId = http2Handler.connection().remote().lastStreamCreated()
+                           + frame.extraStreamIds() * 2;
```

Comment:
I'd probably do `extraStreamIds() == MAX_INT`, and we fix the code to saturate. The `extraStreamIds()` was put there to allow graceful shutdown, although I was expecting something more like `extraStreamIds() == 100` just to put an upper bound on the number of new streams.

Maybe `int extraStreamIds()` should be replaced with `boolean relaxedLastStream()`. If `relaxedLastStream()`, then we use MAX_INT, otherwise we use `lastStreamCreated()`. (Other names I considered are strictLastStream and enforceLastStream.)